services/web/server/src/simcore_service_webserver/director_v2_core_dynamic_services.py

```diff
@@ -0,0 +1,218 @@
+
+from aiohttp import web
+from models_library.projects import ProjectID
+from models_library.services_resources import (
+    ServiceResourcesDict,
+    ServiceResourcesDictHelpers,
+)
+from pydantic.types import PositiveInt
+from servicelib.logging_utils import log_decorator
+from servicelib.utils import logged_gather
+from yarl import URL
+
+from .director_v2_core_base import DataType, request_director_v2
+from .director_v2_exceptions import DirectorServiceError
+from .director_v2_settings import DirectorV2Settings, get_plugin_settings
+
+log = logging.getLogger(__name__)
+
+
+@log_decorator(logger=log)
+async def get_dynamic_services(
+    app: web.Application,
+    user_id: Optional[PositiveInt] = None,
+    project_id: Optional[str] = None,
+) -> list[DataType]:
+    params = {}
+    if user_id:
+        params["user_id"] = user_id
+    if project_id:
+        params["project_id"] = project_id
+
+    settings: DirectorV2Settings = get_plugin_settings(app)
+    backend_url = (settings.base_url / "dynamic_services").update_query(**params)
+
+    services = await request_director_v2(
+        app, "GET", backend_url, expected_status=web.HTTPOk
+    )
+
+    assert isinstance(services, list)  # nosec
+    return services
+
+
+@log_decorator(logger=log)
+async def get_dynamic_service_state(app: web.Application, node_uuid: str) -> DataType:
+    settings: DirectorV2Settings = get_plugin_settings(app)
+    backend_url = settings.base_url / f"dynamic_services/{node_uuid}"
+
+    service_state = await request_director_v2(
+        app, "GET", backend_url, expected_status=web.HTTPOk
+    )
+
+    assert isinstance(service_state, dict)  # nosec
+    return service_state
+
+
+@log_decorator(logger=log)
+async def run_dynamic_service(
+    app: web.Application,
+    user_id: PositiveInt,
+    project_id: str,
+    service_key: str,
+    service_version: str,
+    service_uuid: str,
+    request_dns: str,
+    request_scheme: str,
+    service_resources: ServiceResourcesDict,
+) -> DataType:
+    """
+    Requests to run (i.e. create and start) a dynamic service:
+    - legacy services request is redirected to `director-v0`
+    - dynamic-sidecar `director-v2` will handle the request
+    """
+    data = {
+        "user_id": user_id,
+        "project_id": project_id,
+        "key": service_key,
+        "version": service_version,
+        "node_uuid": service_uuid,
+        "basepath": f"/x/{service_uuid}",
+        "service_resources": ServiceResourcesDictHelpers.create_jsonable(
+            service_resources
+        ),
+    }
+
+    headers = {
+        "X-Dynamic-Sidecar-Request-DNS": request_dns,
+        "X-Dynamic-Sidecar-Request-Scheme": request_scheme,
+    }
+
+    settings: DirectorV2Settings = get_plugin_settings(app)
+    started_service = await request_director_v2(
+        app,
+        "POST",
+        url=settings.base_url / "dynamic_services",
+        data=data,
+        headers=headers,
+        expected_status=web.HTTPCreated,
```

Comment:
Possible future enhancement: Add DIRECTOR_V2_REST_CALL_DEFAULT_REQUEST_TIMEOUT as a configurable settings/env var, place it in all calls that dont already have a custom timeout setting.

Reply:
(This avoids magic numbers, even if they are the library's defaults, like REST Timeouts, that cannot be configured per-deploy in an ad-hoc fashion. Imho devops opinion, all magic numbers should be able to be overwritten as config)

Reply:
pair reviewed